File: python3/find_and_replace_pattern.py
"""
Space : O(len(word))
Time  : O(n * len(word))
"""


class Solution:
    def findAndReplacePattern(self, words: List[str], pattern: str) -> List[str]:
        ans = []
        
        for word in words:
            mem_w = {}
            mem_p = {}
            safe = True
            for i in range(len(word)):
                w = word[i]
                p = pattern[i]
                
                if w not in mem_w and p not in mem_p:
                    mem_w[w] = p
                    mem_p[p] = w
                elif w in mem_w and p in mem_p and mem_w[w] == p and mem_p[p] == w:
                    continue
                else:
                    safe = False
                    break
        
            if not safe:
                continue

            ans.append(word)
        
        
        return ans

File: golang/find_and_replace_pattern.go
package main

/*
Space : O(len(word))
Time  : O(n * len(word))
*/

func findAndReplacePattern(words []string, pattern string) []string {
	var ans []string
	n := len(words)

	for i := 0; i < n; i++ {
		word := words[i]
		mem_w := make(map[byte]byte)
		mem_p := make(map[byte]byte)
		safe := true

		for j := 0; j < len(word); j++ {
			w := word[j]
			p := pattern[j]

			_, okw := mem_w[w]
			_, okp := mem_p[p]

			if !okw && !okp {
				mem_w[w] = p
				mem_p[p] = w
			} else if okw && okp && mem_w[w] == p && mem_p[p] == w {
				continue
			} else {
				safe = false
				break
			}
		}

		if !safe {
			continue
		}
		ans = append(ans, word)

	}

	return ans
}
